File: model/color_algorithm.py
import math
from abc import ABC, abstractmethod
from hashlib import sha256

from model.color_memo import ColorMemo
from model.rgb import RGB

RGB_SCALAR: int = 192
RGB_OFFSET: int = 128


def hash(val: str) -> str:
    return sha256(val.encode("utf-8")).hexdigest()


class ColorAlgorithm(ABC):
    lookup_key: str
    num_buckets: int
    scale: float
    reverse: bool

    @abstractmethod
    def evaluate(self, percent: float, idx: int, total_leds: int) -> RGB:
        pass

    @abstractmethod
    def is_linear(self) -> bool:
        # Return if the color algorithm can be represented as a 1D array, purely linear motion
        pass

    def is_reverse(self) -> bool:
        return self.reverse

    def get_bucket(self, percent: float) -> int:
        return math.floor(percent * self.num_buckets)

    def set_adjustment_level(self, level: int) -> None:
        pass


class RainbowRGB(ColorAlgorithm):
    def __init__(
        self,
        offset: float,
        color_memo: ColorMemo,
        scale: float = 1.0,
        reverse: bool = False,
    ):
        self._offset = offset
        self._memo = color_memo
        self.num_buckets = 50
        self.scale = scale
        self.reverse = reverse
        self.lookup_key = self._calculate_lookup_key()

    def _calculate_lookup_key(self) -> str:
        return hash(self.__class__.__name__ + f"sc_{self.scale}")

    def evaluate(self, percent: float, _: int, __: int) -> RGB:
        offset_percent = (percent + self._offset) % 1.0
        bucket = self.get_bucket(offset_percent)

        precomputed = self._memo.get(self.lookup_key, bucket)
        if precomputed:
            return precomputed

        a = offset_percent * 2 * math.pi
        r = math.sin(a) * RGB_SCALAR + RGB_OFFSET
        g = math.sin(a - (2 * math.pi / 3)) * RGB_SCALAR + RGB_OFFSET
        b = math.sin(a - (4 * math.pi / 3)) * RGB_SCALAR + RGB_OFFSET

        rgb = RGB(r, g, b)
        self._memo.set_value(self.lookup_key, bucket, rgb)
        return rgb

    def is_linear(self):
        return True


class Comet(ColorAlgorithm):
    def __init__(
        self,
        offset: float,
        color_memo: ColorMemo,
        scale: float = 1.0,
        reverse: bool = False,
        color_offset: float = 0.0,
    ):
        self._offset = offset
        self._memo = color_memo
        self.num_buckets = 200
        self.scale = scale
        self.reverse = reverse
        self._color_offset = color_offset
        self.adjustment_level = 0
        self.lookup_key = self._calculate_lookup_key()

    def _calculate_lookup_key(self) -> str:
        return hash(
            self.__class__.__name__
            + f"sc_{self.scale}"
            + f"co_{str(self._color_offset)}"
            + f"al_{str(self.adjustment_level)}"
        )

    def set_adjustment_level(self, level: int) -> None:
        self.adjustment_level = level
        self.lookup_key = self._calculate_lookup_key()

    def evaluate(self, percent: float, _: int, __: int) -> RGB:
        offset_percent = (percent + self._offset) % 1.0
        bucket = self.get_bucket(offset_percent)
        precomputed = self._memo.get(self.lookup_key, bucket)
        if precomputed:
            return precomputed

        dot_count = 3
        count_per_grouping = self.num_buckets / dot_count
        dropoff_factor = 1 / (count_per_grouping * 2 / 3)
        mod = bucket % (self.num_buckets / dot_count)
        intensity = max(1 - (mod * dropoff_factor), 0)

        a = (
            offset_percent * 2 * math.pi
            + self._color_offset
            + (self.adjustment_level / 30.0)
        )
        r = intensity * (math.sin(a) * RGB_SCALAR + RGB_OFFSET)
        g = intensity * (math.sin(a - (2 * math.pi / 3)) * RGB_SCALAR + RGB_OFFSET)
        b = intensity * (math.sin(a - (4 * math.pi / 3)) * RGB_SCALAR + RGB_OFFSET)

        rgb = RGB(r, g, b)
        self._memo.set_value(self.lookup_key, bucket, rgb)
        return rgb

    def is_linear(self):
        return True


class PurpleGreenOrangeComet(Comet):
    def __init__(
        self,
        offset: float,
        color_memo: ColorMemo,
        scale: float = 1.0,
        reverse: bool = False,
    ):
        pgo_color_offset = 2 * math.pi / 3
        super().__init__(offset, color_memo, scale, reverse, pgo_color_offset)


class PastelRGB(ColorAlgorithm):
    def __init__(
        self,
        offset: float,
        color_memo: ColorMemo,
        red_scalar: float = 110,
        red_offset: float = 145,
        green_scalar: float = 110,
        green_offset: float = 145,
        blue_scalar: float = 110,
        blue_offset: float = 145,
        scale: float = 1.0,
        reverse=False,
    ):
        self._offset = offset
        self._memo = color_memo
        self._red_scalar = red_scalar
        self._red_offset = red_offset
        self._green_scalar = green_scalar
        self._green_offset = green_offset
        self._blue_scalar = blue_scalar
        self._blue_offset = blue_offset
        self.scale = scale
        self.lookup_key = hash(
            self.__class__.__name__
            + f"rs_{red_scalar}"
            + f"ro_{red_offset}"
            + f"gs_{green_scalar}"
            + f"go_{green_offset}"
            + f"bs_{blue_scalar}"
            + f"bo_{blue_offset}"
            + f"sc_{self.scale}"
        )
        self.num_buckets = 50
        self.reverse = reverse

    def evaluate(self, percent: float, _: int, __: int) -> RGB:
        offset_percent = (percent + self._offset) % 1.0
        bucket = self.get_bucket(offset_percent)
        precomputed = self._memo.get(self.lookup_key, bucket)
        if precomputed:
            return precomputed

        a = offset_percent * 2 * math.pi
        r = self._red_scalar * (math.sin(a) + 1) + self._red_offset
        g = self._green_scalar * (math.sin(a - (2 * math.pi / 3))) + self._green_offset
        b = self._blue_scalar * (math.sin(a - (4 * math.pi / 3))) + self._blue_offset

        rgb = RGB(r, g, b)
        self._memo.set_value(self.lookup_key, bucket, rgb)
        return rgb

    def is_linear(self):
        return True


class Yoyo(ColorAlgorithm):
    def __init__(
        self,
        offset: float,
        color_memo: ColorMemo,
        scale: float = 1.0,
        reverse: bool = False,
        color_offset: float = 0.0,
    ):
        self._offset = offset
        self._memo = color_memo
        self.num_buckets = 200
        self.scale = scale
        self.reverse = reverse
        self._color_offset = color_offset
        self.adjustment_level = 0
        self.lookup_key = self._calculate_lookup_key()

    def _calculate_lookup_key(self) -> str:
        return hash(
            self.__class__.__name__
            + f"sc_{self.scale}"
            + f"co_{str(self._color_offset)}"
            + f"al_{str(self.adjustment_level)}"
        )

    def set_adjustment_level(self, level: int) -> None:
        self.adjustment_level = level
        self.lookup_key = self._calculate_lookup_key()

    def evaluate(self, percent: float, idx: int, total_count: int) -> RGB:
        # 2.0 is for the yoyo effect
        original_offset_percent = percent + self._offset
        offset_percent = abs((2 * original_offset_percent) % 2.0)
        reverse = False
        if offset_percent >= 1.0:
            offset_percent = 2 - offset_percent
            reverse = True

        bucket = self.get_bucket(original_offset_percent)
        full_key = f"{idx}_{bucket}"
        precomputed = self._memo.get(self.lookup_key, full_key)
        if precomputed:
            return precomputed

        # less dropoff in the middle
        tail_length_percent = math.sin(offset_percent * math.pi) / 4.0
        tail_length_count = tail_length_percent * total_count

        intensity = 0
        if reverse:
            head_idx = math.floor(offset_percent * total_count)
            if head_idx + tail_length_count >= idx and idx >= head_idx:
                intensity = 1
        else:
            head_idx = math.floor(offset_percent * total_count)
            if head_idx - tail_length_count <= idx and idx <= head_idx:
                intensity = 1

        # for each bucket, use the tail length to determine if the specific idx should be displayed
        # dropoff_factor = 1 / (count_per_grouping * diff)

        # intensity = max(1 - (bucket * dropoff_factor), 0)
        # intensity = diff

        # white
        r = intensity * 255
        g = intensity * 255
        b = intensity * 255

        # highlight head in red
        # if idx == head_idx:
        #     r = 255
        #     g = 0
        #     b = 0
        #     # print(
        #     #     f"offset_percent: {round(offset_percent, 2)}, tail_length_percent: {round(tail_length_percent, 2)}, tail length count: {round(tail_length_count, 2)}"
        #     # )

        rgb = RGB(r, g, b)
        self._memo.set_value(self.lookup_key, full_key, rgb)
        return rgb

    def is_linear(self):
        return False

File: cpp/src/color_algorithm.cpp
#define _USE_MATH_DEFINES
#include "../include/color_algorithm.h"

#include <math.h>

#include <algorithm>
#include <cmath>
#include <vector>

#include "../include/color_map.h"
#include "../include/rgb.h"
#include "../include/util.h"

const float POWER_SCALE = 0.07;  // 7% of max power

ColorMap* ColorAlgorithm::initialize_linear_color_map() {
  ColorMap* color_map = new LinearColorMap(bucket_size_);
  for (int i = 0; i < bucket_size_; i++) {
    const float percent = float(i) / bucket_size_;
    const RGB rgb = compute(percent, i, bucket_size_);
    color_map->add_color(i, i, bucket_size_, rgb);
  }
  return color_map;
}

ColorMap* ColorAlgorithm::initialize_full_color_map(const int leds_per_strip) {
  ColorMap* color_map = new FullColorMap(bucket_size_);
  for (int i = 0; i < bucket_size_; i++) {
    const float percent = float(i) / bucket_size_;
    for (int j = 0; j < bucket_size_; j++) {
      const RGB rgb = compute(percent, j, leds_per_strip);
      color_map->add_color(i, j, leds_per_strip, rgb);
    }
  }
  return color_map;
}

void ColorAlgorithm::reset_color_map() { color_map_ = NULL; }

Rainbow::Rainbow(const int bucket_size) { bucket_size_ = bucket_size; }

ColorMap* Rainbow::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_linear_color_map());
  }
  return color_map_.get();
}

RGB Rainbow::compute(float percent, int idx, int total_idx) const {
  const int RGB_SCALAR = 128;
  const int RGB_OFFSET = 128;

  float a = percent * 2 * M_PI;
  const float r_sin = sin(a);
  const float g_sin = sin(a - (2 * M_PI / 3));
  const float b_sin = sin(a - (4 * M_PI / 3));
  float r = POWER_SCALE * (r_sin * RGB_SCALAR + RGB_OFFSET);
  float g = POWER_SCALE * (g_sin * RGB_SCALAR + RGB_OFFSET);
  float b = POWER_SCALE * (b_sin * RGB_SCALAR + RGB_OFFSET);
  return RGB(r, g, b);
}

bool Rainbow::is_linear() const { return true; }

Pastel::Pastel(const int bucket_size, int red_scalar, int red_offset, int green_scalar,
               int green_offset, int blue_scalar, int blue_offset) {
  bucket_size_ = bucket_size;
  red_scalar_ = red_scalar;
  red_offset_ = red_offset;
  green_scalar_ = green_scalar;
  green_offset_ = green_offset;
  blue_scalar_ = blue_scalar;
  blue_offset_ = blue_offset;
}

ColorMap* Pastel::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_linear_color_map());
  }
  return color_map_.get();
}

RGB Pastel::compute(float percent, int idx, int total_idx) const {
  float a = percent * 2 * M_PI;
  float r_sin = sin(a);
  float g_sin = sin(a - (2 * M_PI / 3));
  float b_sin = sin(a - (4 * M_PI / 3));

  float r = POWER_SCALE * (r_sin * red_scalar_ + red_offset_);
  float g = POWER_SCALE * (g_sin * green_scalar_ + green_offset_);
  float b = POWER_SCALE * (b_sin * blue_scalar_ + blue_offset_);
  return RGB(r, g, b);
}
bool Pastel::is_linear() const { return true; }

Comet::Comet(const int bucket_size) { bucket_size_ = bucket_size; }

ColorMap* Comet::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_linear_color_map());
  }
  return color_map_.get();
}

RGB Comet::compute(float percent, int idx, int total_idx) const {
  const int RGB_SCALAR = 128;
  const int RGB_OFFSET = 128;

  // Split into groups and determine the relative intensity
  const int dot_count = 6;
  const int count_per_grouping = float(bucket_size_) / dot_count;
  const float dropoff_factor = 1.0 / (count_per_grouping * 2 / 3);
  const int bucket = int(round(percent * bucket_size_));
  const int mod = bucket % (int)(float(bucket_size_) / dot_count);
  const float intensity = std::max(1 - (mod * dropoff_factor), 0.0f);

  // Figure out the colors
  const float a = percent * M_PI * 2;
  const float r_sin = sin(a);
  const float g_sin = sin(a - (2 * M_PI / 3));
  const float b_sin = sin(a - (4 * M_PI / 3));
  const float r = POWER_SCALE * intensity * (r_sin * RGB_SCALAR + RGB_OFFSET);
  const float g = POWER_SCALE * intensity * (g_sin * RGB_SCALAR + RGB_OFFSET);
  const float b = POWER_SCALE * intensity * (b_sin * RGB_SCALAR + RGB_OFFSET);
  return RGB(r, g, b);
}

bool Comet::is_linear() const { return true; }

Yoyo::Yoyo(const int bucket_size, const int leds_per_strip, const int r_scalar, const int g_scalar,
           const int b_scalar)
    : leds_per_strip_(leds_per_strip), time_scale_(3.0) {
  bucket_size_ = bucket_size;
  r_scalar_ = r_scalar;
  g_scalar_ = g_scalar;
  b_scalar_ = b_scalar;
}

ColorMap* Yoyo::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_full_color_map(leds_per_strip_));
  }
  return color_map_.get();
}

RGB Yoyo::compute(float input_percent, int idx, int total_idx) const {
  float percent = fmod(input_percent * time_scale_, 1.0);
  // 2.0 is for the yoyo effect
  float offset_percent = percent * 2.0;
  bool reverse = false;
  if (offset_percent >= 1.0) {
    offset_percent = 2.0 - offset_percent;
    reverse = true;
  }

  // S curve for the head idx to simulate different speeds
  const float min_head_percent = 0.05;
  const float comet_range_percent = 0.9;
  const float min_head_idx = min_head_percent * total_idx;
  const int head_idx =
      std::round(s_curve(offset_percent) * total_idx * comet_range_percent + min_head_idx);

  // use a normal distribution to determine the tail length
  const float tail_length_percent = pow(gaussian_value(offset_percent), 1.5);
  const float tail_length_count = tail_length_percent * total_idx;

  float intensity = 0.0f;
  int offset_from_head = head_idx - idx;
  if (reverse) {
    offset_from_head = idx - head_idx;
  }
  if (offset_from_head >= 0 && offset_from_head <= tail_length_count) {
    intensity = 1 * pow(0.9, offset_from_head);
  }

  // give the head a bit extra
  const int pre_head_count = 4;
  if (offset_from_head <= 0 && offset_from_head > -1.0 * pre_head_count) {
    intensity = 1;
  }

  // adjust up due to the relative darkness of custom colors
  const float additional_scale_factor = 255.0 * 3 / (r_scalar_ + g_scalar_ + b_scalar_);

  float r = r_scalar_ * intensity * POWER_SCALE * additional_scale_factor;
  float g = g_scalar_ * intensity * POWER_SCALE * additional_scale_factor;
  float b = b_scalar_ * intensity * POWER_SCALE * additional_scale_factor;

  // highlight the head red
  // if (idx == head_idx) {
  //   r = 255 * POWER_SCALE;
  //   g = 0;
  //   b = 0;
  // }

  return RGB(r, g, b);
}
bool Yoyo::is_linear() const { return false; }

Wubwub::Wubwub(const int bucket_size, const int leds_per_strip,
               const std::vector<ColorPoint> colors)
    : color_gradient_(ColorGradient(colors)), leds_per_strip_(leds_per_strip) {
  bucket_size_ = bucket_size;
}

ColorMap* Wubwub::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_full_color_map(leds_per_strip_));
  }
  return color_map_.get();
}

RGB Wubwub::compute(float percent, int idx, int total_idx) const {
  const float wave_count = 3.0;

  const float idx_percent = float(idx) / total_idx;
  const float wave1_shape = (1.0 * cos(wave_count * M_PI * 2 * idx_percent) + 1) / 2.0;

  // Rotate the wave around 0.5
  float wave_value = 0;
  if (wave1_shape >= 0.5 && percent <= 0.5) {
    wave_value = wave1_shape - 4.0 * percent * abs(wave1_shape - 0.5);
  }
  if (wave1_shape >= 0.5 && percent > 0.5) {
    wave_value = wave1_shape - 4.0 * (1.0 - percent) * abs(wave1_shape - 0.5);
  }
  if (wave1_shape < 0.5 && percent <= 0.5) {
    wave_value = wave1_shape + 4.0 * percent * abs(wave1_shape - 0.5);
  }
  if (wave1_shape < 0.5 && percent > 0.5) {
    wave_value = wave1_shape + 4.0 * (1.0 - percent) * abs(wave1_shape - 0.5);
  }

  float wave_amp = wave_value;

  const RGB gradient_rgb = color_gradient_.getColorAtValue(wave_amp);
  const float r = gradient_rgb.red * POWER_SCALE;
  const float g = gradient_rgb.green * POWER_SCALE;
  const float b = gradient_rgb.blue * POWER_SCALE;

  return RGB(r, g, b);
}

bool Wubwub::is_linear() const { return false; }

Sparkshot::Sparkshot(const int bucket_size, const int leds_per_strip, const bool vary_intensity,
                     const std::vector<Shot> shots)
    : shots_(shots), vary_intensity_(vary_intensity) {
  bucket_size_ = bucket_size;
  leds_per_strip_ = leds_per_strip;
}

ColorMap* Sparkshot::get_color_map() {
  if (color_map_ == NULL) {
    color_map_ = std::unique_ptr<ColorMap>(initialize_full_color_map(leds_per_strip_));
  }
  return color_map_.get();
}

RGB Sparkshot::compute(float percent, int idx, int total_idx) const {
  const float min_intensity = 0.2;
  RGB rgb = RGB(0, 0, 0);

  // Iterate over shots backwards (last shot gets precedence)
  for (auto shot = shots_.rbegin(); shot != shots_.rend(); ++shot) {
    const float start_idx_pct = shot->start_idx_pct_;
    const float end_idx_pct = shot->end_idx_pct_;
    const float dropoff_factor = shot->dropoff_factor_;
    const RGB color = shot->color_;

    // derived
    const int start_idx = std::round(start_idx_pct * total_idx);
    const int end_idx = std::round(end_idx_pct * total_idx);
    const int total_distance = abs(end_idx - start_idx);
    const bool forward = end_idx > start_idx;
    const float distance = percent * total_distance;
    const int head_idx =
        forward ? std::round(start_idx + distance) : std::round(start_idx - distance);
    const int distance_from_head = abs(head_idx - idx);
    float intensity = pow(dropoff_factor, distance_from_head);

    // Move to the next shot if this one isn't bright enough
    if (intensity < min_intensity) {
      continue;
    }

    if (!vary_intensity_) {
      intensity = 1;
    }

    // Figure out the colors
    const float additional_scale_factor = 255.0 * 3 / (color.red + color.green + color.blue);
    rgb = RGB(POWER_SCALE * intensity * color.red * additional_scale_factor,
              POWER_SCALE * intensity * color.green * additional_scale_factor,
              POWER_SCALE * intensity * color.blue * additional_scale_factor);

    // If we found a color, drop out
    break;
  }
  return rgb;
}

bool Sparkshot::is_linear() const { return false; }
